mlx/linalg.cpp

```diff
@@ -292,4 +292,41 @@ array cholesky(
       {a});
 }
 
+array pinv(const array& a, StreamOrDevice s /* = {} */) {
+  if (a.dtype() != float32) {
+    std::ostringstream msg;
+    msg << "[linalg::pinv] Arrays must type float32. Received array "
+        << "with type " << a.dtype() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+  if (a.ndim() < 2) {
+    std::ostringstream msg;
+    msg << "[linalg::pinv] Arrays must have >= 2 dimensions. Received array "
+           "with "
+        << a.ndim() << " dimensions.";
+    throw std::invalid_argument(msg.str());
+  }
+
+  const auto m = a.shape(-2);
+  const auto n = a.shape(-1);
+  const auto k = std::min(m, n);
+  const auto rank = a.ndim();
+
+  auto outs = linalg::svd(a, s);
+  auto U = outs[0];
+  auto S = outs[1];
+  auto Vt = outs[2];
+
+  std::vector<int> pinv_shape;
+  if (m <= n) {
+    pinv_shape = {n, k};
+  } else {
+    pinv_shape = {k, m};
+  }
+  array pinv(pinv_shape, a.dtype(), nullptr, {});
+  const auto U_slice = slice(U, {0, 0}, {m, k});
+  const auto Vt_slice = slice(Vt, {0, 0}, {k, n});
+  return matmul(matmul(transpose(Vt_slice), diag(1.0 / S)), transpose(U_slice));
```

Comment:
No need for two matmuls and a `diag` this can be done with a division. Also the streams should be appropriately used.

```cpp
std::vector<int> starts(a.ndim(), 0);
std::vector<int> ends = a.shape();
int i = a.ndim()-2;
int j = a.ndim()-1;
ends[i] = n;
ends[j] = k;
array U_slice = slice(U, starts, ends, s);
ends[i] = k;
ends[j] = m;
array V_slice = slice(Vt, starts, ends, s);
return swapaxes(matmul(divide(U_slice, expand_dims(S, -2, s), s), V_slice, s), -1, -2, s);
```

Reply:
Updated ✅ - in new commit 587a30f6; [diff here](https://github.com/ml-explore/mlx/pull/875/commits/587a30f6bd32a12a0f58ed660fc7a5144820c3b1#diff-9f87ac95c2fab89844d98cd9bd7727e2e7a872880ea49db8baf529bb43bdfcddR328)

Btw, I think your message above mixed up the `ends[]` dims for `U_slice` vs `V_slice` leading to their dims being `{n, k}` and `{k, m}` respectively.  But IIUC they should instead be `U_slice -> {m, k}` and `V_slice -> {k, n}`. The _exact_ code you shared above caused a test error and after investigating I updated the code to use ensure dims of `U_slice` is `{m, k}` and `V_slice` is `{k, n}`.  I verified the results comparing to numpy/matlab. But please double check my changes. Thanks!